### PR TITLE
feat: introduce puppeteer/Errors

### LINKED ISSUE #1694: Conditional catch for errors
<sup>Issue template didn't really apply.</sup>

I can't find anything in the api docs page about error handling. The main type of error that's encountered is a timeout. The issue is that there's no documented way to distinguish the type of error received in the `catch` after awaiting a promise that rejects.

The issue is that good code (most of the time) needs to do conditional catches to avoid silencing unexpected errors. There doesn't seem to be a good, stable way to do this currently.

I tried inspecting some of the errors. For these tests, the output is logging `error` and then logging `{...error}`.

```js
await page.click('doesntexist');

// Result
{ AssertionError [ERR_ASSERTION]: No node found for selector: doesntexist
    at Console.assert (console.js:188:23)
    at Page.click (/private/tmp/temp-puppet-7695/node_modules/puppeteer/lib/Page.js:770:13)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  generatedMessage: false,
  name: 'AssertionError [ERR_ASSERTION]',
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '==' }
{ generatedMessage: false,
  name: 'AssertionError [ERR_ASSERTION]',
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '==' }
```

Gives a generic `AssertionError` with no information about the cause, except if you want to parse the error message, which seems very unstable as it can change without much notice.

The next issue is `waitFor`:

```js
await page.waitFor('doesntexist');

// Result
Error: waiting failed: timeout 30000ms exceeded
    at Timeout.WaitTask._timeoutTimer.setTimeout (/private/tmp/temp-puppet-7695/node_modules/puppeteer/lib/FrameManager.js:593:58)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
{}
```

Again, the only way I see to extract information is the error message.

---

There are a few potential solutions here:

1. document the error messages and give an example regex to check them. This requires no code changes, but makes it clear that the error messages changing would be a breaking change. 
2. don't change the errors, but give official utility functions like `puppeteer.isTimeoutError(error)`. Very little code changes, but the api will be stable across versions. Better than 1.
3. make each error either an Error subclass, or give it properties e.g. `type: 'PuppeteerTimeout', selector: 'doesntexist'`. This is about tied with 2, since it gives a readable way to check the error type.

Thoughts? Sorry if this has come up before, but I couldn't find anything relevant.

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch adds a new require, `puppeteer/Errors`, that
holds all the Puppeteer-specific error classes.

Currently, the only custom error class we use is `TimeoutError`. We'll
expand in future with `CrashError` and some others.

Fixes #1694.

### CODE REVIEW

File: utils/doclint/check_public_api/index.js [hunk 1]
```diff
@@ -42,6 +43,10 @@ const EXCLUDE_METHODS = new Set([
   'JSHandle.toString',
 ]);
 
+const EXCLUDE_PROPERTIES = new Set([
```
Comment (JoelEinbinder) [inline utils/doclint/check_public_api/index.js]: Methods are properties. I think you can just rename EXCLUDE_METHODS to EXCLUDE_PROPERTIES and add `TimeoutError.name` into it. They can't conflict.

**Review by JoelEinbinder — APPROVED** [review]

File: utils/doclint/check_public_api/index.js (hunk 1 above)
Comment (aslushnikov) [reply utils/doclint/check_public_api/index.js]: Done.

File: docs/api.md [hunk 2]
```diff
@@ -316,6 +318,34 @@ If puppeteer doesn't find them in environment, lowercased variant of these varia
...
+An example of handling timeout error:
+```js
+const {TimeoutError} = require('puppeteer/Errors');
+
+// ...
+
+try {
+  await page.waitForSelector('.foo');
+} catch (e) {
+  if (e instanceof TimeoutError) {
```
Comment (felixfbecker) [inline docs/api.md:342]: Imo this is not a good suggestion. `instanceof` is an antipattern. Assert against interfaces, not implementations. Better would be `if (e.name === 'TimeoutError')`

File: lib/Errors.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class CustomError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
```
Comment (felixfbecker) [inline lib/Errors.js:20]: This will break if the code gets minified in some way. It is better to assign a string constant in each of the subclasses.

Comment (aslushnikov) [reply lib/Errors.js:20]: > This will break if the code gets minified in some way. It is better to assign a string constant in each of the subclasses.

We don't assume code minification/transpilation usecases, and we already fail miserably with some tools like babel. Does it hurt you? If yes, please file an issue so that your concerns are tracked.

File: docs/api.md (hunk 2 above)
Comment (aslushnikov) [reply docs/api.md:342]: > Imo this is not a good suggestion. instanceof is an antipattern. Assert against interfaces, not implementations. Better would be if (e.name === 'TimeoutError')

I'm not sure I follow. Can you please expand on this?

Comment (felixfbecker) [reply docs/api.md:342]: When using `instanceof`, you check if an object is an instance of a concrete class implementation. It is better to check if the object adheres to the interface you expect, e.g. if it has the needed properties or in this case, if the error has a certain error code. This is the approach taken by NodeJS errors (you cannot even check with `instanceof` as the classes are not exposed, just check `error.code`) and also browser APIs (for example, `AbortError` is not even a class, but you always have to check `err.name === 'AbortError'`).

Imagine you had two versions of Puppeteer in your tree, because some library depends on it, but another library or your root package does too, but a different version (this might have happened silently through an update). Now when you import `TimeoutError` from Puppeteer, the class will actually be a different class, and all your `instanceof` checks will fail - even though it clearly is a `TimeoutError`, and says so in `error.name`. Relying on a certain layout in node_modules in brittle.

This is why it's best practice to avoid `instanceof` and check error names/codes instead.

Comment (aslushnikov) [reply docs/api.md:342]: > Imagine you had two versions of Puppeteer in your tree, because some library depends on it, but another library or your root package does too, but a different version (this might have happened silently through an update).

Let me re-iterate if understand correctly. I'm developing project `myproj` that depends on `Puppeteer` and also has a dependency on `FancyLib` that happens to use Puppeteer as well. In addition, in order to make things extreme (and prevent tree flattening), let's consider these Puppeteers to be of different versions.

```
myproj
|
|--FancyLib
|      |--Puppeteer v1.7.0
|
|--Puppeteer v1.8.0
```

In this setup, checking errors emitted by FancyLib to be Puppeteer errors is a logical error since they're not part of FancyLib API:

```js
import {TimeoutError} from 'puppeteer/Errors';
import {FancyLib} from 'fancylib';
try {
  await FancyLib.doStuff();
} catch(e) {
  // Logic error: FancyLib has nothing to do with Puppeteer.
  if (e instanceof TimeoutError) {
  }
}
```

Instead, if FancyLib wants to provide error handling as a part of its api, it should provide explicit errors for its clients to catch:
```js
import {FancyLib, TimeoutError} from 'fancylib';
try {
  await FancyLib.doStuff();
} catch(e) {
  // Correct: FancyLib exposes TimeoutError as part of its API
  if (e instanceof TimeoutError) {
  }
}
```

So when using puppeteer library directly, its errors are part of its api, so its safe to instanceof against them. Using event names as an alternative to classes is possible, but we'd have to come up with a very distinct error names so that there are less chances for them to clash with other libraries. 

To sum up, both approaches are viable, but I'd prefer to keep the doc as-is.

Comment (felixfbecker) [reply docs/api.md:342]: > we'd have to come up with a very distinct error names so that there are less chances for them to clash with other libraries.

This is the point - you _want_ to have them clash. What else could a `TimeoutError` be other than an error because of a timeout? I.e. as a user, you don't care if it's a timeout error from library FooBarLib, you only care if it is a timeout error. Just like I don't care if it is a FooBarLibAbortError, I only care if it is an AbortError.

Yes, both approaches work usually. I've just seen people being tripped up too often because an update changed their node_modules and suddenly their `instanceof` silently failed that worked before, so I believe it's better to spread the practice of checking against names.

If you don't want to change it that's fine, it was just a suggestion.

Comment (aslushnikov) [reply docs/api.md:342]: > This is the point - you want to have them clash. What else could a TimeoutError be other than an error because of a timeout?

That's fair for the timeout error, but I disagree in general. For example, the `CrashError` we're about to introduce for target crashes might mean very different things from library to library, with different recovery strategies.

IMO since node.js platform doesn't define a set of common errors with a well-defined semantics, we shouldn't invent one as well.

> If you don't want to change it that's fine, it was just a suggestion.

Thank you for the suggestion!

Comment (felixfbecker) [reply docs/api.md:342]: I can definitely see your point.
